File: backend/src/services/person_service.py
from sqlalchemy.orm import Session
from sqlalchemy import select, and_
from typing import List, Optional
from cryptography.fernet import Fernet
import os
from ..models.person import Person, person_relationships
from ..schemas.person import PersonCreate, PersonUpdate, PersonRelationshipCreate


class PersonService:
    """Service for managing people and their relationships"""
    
    def __init__(self):
        # Initialize encryption key for SSN
        # In production, this should be stored securely (e.g., environment variable, secrets manager)
        encryption_key = os.getenv('SSN_ENCRYPTION_KEY')
        if not encryption_key:
            # Generate a key if not provided (for development only)
            encryption_key = Fernet.generate_key()
        if isinstance(encryption_key, str):
            encryption_key = encryption_key.encode()
        self.cipher = Fernet(encryption_key)
    
    def _encrypt_ssn(self, ssn: str) -> str:
        """Encrypt a Social Security Number"""
        if not ssn:
            return None
        return self.cipher.encrypt(ssn.encode()).decode()
    
    def _decrypt_ssn(self, encrypted_ssn: str) -> str:
        """Decrypt a Social Security Number"""
        if not encrypted_ssn:
            return None
        return self.cipher.decrypt(encrypted_ssn.encode()).decode()
    
    def create_person(self, db: Session, person_data: PersonCreate) -> Person:
        """Create a new person"""
        # Encrypt SSN if provided
        ssn_encrypted = None
        if person_data.ssn:
            ssn_encrypted = self._encrypt_ssn(person_data.ssn)
        
        # Create person without SSN in the data dict
        person_dict = person_data.model_dump(exclude={'ssn'})
        person = Person(**person_dict, ssn_encrypted=ssn_encrypted)
        
        db.add(person)
        db.commit()
        db.refresh(person)
        return person
    
    def get_person(self, db: Session, person_id: int) -> Optional[Person]:
        """Get a person by ID"""
        return db.query(Person).filter(Person.id == person_id).first()
    
    def get_people(
        self,
        db: Session,
        skip: int = 0,
        limit: int = 100,
        include_inactive: bool = False,
        search: Optional[str] = None
    ) -> tuple[List[Person], int]:
        """Get a list of people with pagination and optional search"""
        query = db.query(Person)
        
        if not include_inactive:
            query = query.filter(Person.is_active == 1)
        
        if search:
            search_filter = f"%{search}%"
            query = query.filter(
                (Person.name.ilike(search_filter)) |
                (Person.email.ilike(search_filter)) |
                (Person.employer.ilike(search_filter))
            )
        
        total = query.count()
        people = query.offset(skip).limit(limit).all()
        
        return people, total
    
    def update_person(
        self,
        db: Session,
        person_id: int,
        person_data: PersonUpdate
    ) -> Optional[Person]:
        """Update a person"""
        person = self.get_person(db, person_id)
        if not person:
            return None
        
        update_data = person_data.model_dump(exclude_unset=True, exclude={'ssn'})
        
        # Handle SSN encryption if provided
        if person_data.ssn is not None:
            update_data['ssn_encrypted'] = self._encrypt_ssn(person_data.ssn)
        
        for field, value in update_data.items():
            setattr(person, field, value)
        
        db.commit()
        db.refresh(person)
        return person
    
    def delete_person(self, db: Session, person_id: int) -> bool:
        """Soft delete a person"""
        person = self.get_person(db, person_id)
        if not person:
            return False
        
        person.is_active = 0
        db.commit()
        return True
    
    def add_relationship(
        self,
        db: Session,
        person_id: int,
        relationship_data: PersonRelationshipCreate
    ) -> bool:
        """Add a relationship between two people"""
        # Verify both people exist
        person = self.get_person(db, person_id)
        related_person = self.get_person(db, relationship_data.related_person_id)
        
        if not person or not related_person:
            return False
        
        # Check if relationship already exists
        existing = db.execute(
            select(person_relationships).where(
                and_(
                    person_relationships.c.person_id == person_id,
                    person_relationships.c.related_person_id == relationship_data.related_person_id
                )
            )
        ).first()
        
        if existing:
            return False
        
        # Insert relationship
        db.execute(
            person_relationships.insert().values(
                person_id=person_id,
                related_person_id=relationship_data.related_person_id,
                relationship_type=relationship_data.relationship_type
            )
        )
        db.commit()
        return True
    
    def remove_relationship(
        self,
        db: Session,
        person_id: int,
        related_person_id: int
    ) -> bool:
        """Remove a relationship between two people"""
        result = db.execute(
            person_relationships.delete().where(
                and_(
                    person_relationships.c.person_id == person_id,
                    person_relationships.c.related_person_id == related_person_id
                )
            )
        )
        db.commit()
        return result.rowcount > 0
    
    def get_relationships(
        self,
        db: Session,
        person_id: int
    ) -> List[dict]:
        """Get all relationships for a person"""
        relationships = db.execute(
            select(person_relationships).where(
                person_relationships.c.person_id == person_id
            )
        ).all()
        
        return [
            {
                'person_id': rel.person_id,
                'related_person_id': rel.related_person_id,
                'relationship_type': rel.relationship_type,
                'created_at': rel.created_at,
                'updated_at': rel.updated_at
            }
            for rel in relationships
        ]


# Singleton instance
person_service = PersonService()

File: go-backend/internal/services/person_service.go
package services

import (
	"encoding/json"
	"errors"
	"time"

	"github.com/estate-docs/go-backend/internal/models"
	"gorm.io/gorm"
)

type PersonService struct {
	DB *gorm.DB
}

func NewPersonService(db *gorm.DB) *PersonService {
	return &PersonService{DB: db}
}

type PersonCreateInput struct {
	Name                     string           `json:"name"`
	PhoneNumber              *string          `json:"phone_number"`
	DateOfBirth              *string          `json:"date_of_birth"`
	SSN                      *string          `json:"ssn"`
	Email                    *string          `json:"email"`
	Employer                 *string          `json:"employer"`
	Occupation               *string          `json:"occupation"`
	MailingAddress           *json.RawMessage `json:"mailing_address"`
	PhysicalAddress          *json.RawMessage `json:"physical_address"`
	TrustorIsLiving          *int             `json:"trustor_is_living"`
	DateOfDeath              *string          `json:"date_of_death"`
	TrustorDeathCertReceived *int             `json:"trustor_death_certificate_received"`
	TrustorOfSoundMind       *int             `json:"trustor_of_sound_mind"`
	TrustorHasRelinquished   *int             `json:"trustor_has_relinquished"`
	TrustorRelinquishedDate  *string          `json:"trustor_relinquished_date"`
	TrustorRelingDocReceived *int             `json:"trustor_reling_doc_received"`
}

type PersonUpdateInput struct {
	Name                     *string          `json:"name"`
	PhoneNumber              *string          `json:"phone_number"`
	DateOfBirth              *string          `json:"date_of_birth"`
	SSN                      *string          `json:"ssn"`
	Email                    *string          `json:"email"`
	Employer                 *string          `json:"employer"`
	Occupation               *string          `json:"occupation"`
	MailingAddress           *json.RawMessage `json:"mailing_address"`
	PhysicalAddress          *json.RawMessage `json:"physical_address"`
	TrustorIsLiving          *int             `json:"trustor_is_living"`
	DateOfDeath              *string          `json:"date_of_death"`
	TrustorDeathCertReceived *int             `json:"trustor_death_certificate_received"`
	TrustorOfSoundMind       *int             `json:"trustor_of_sound_mind"`
	TrustorHasRelinquished   *int             `json:"trustor_has_relinquished"`
	TrustorRelinquishedDate  *string          `json:"trustor_relinquished_date"`
	TrustorRelingDocReceived *int             `json:"trustor_reling_doc_received"`
}

type PersonRelationshipInput struct {
	RelatedPersonID  int    `json:"related_person_id"`
	RelationshipType string `json:"relationship_type"`
}

func (s *PersonService) Create(input PersonCreateInput) (*models.Person, error) {
	p := models.Person{
		Name:                     input.Name,
		PhoneNumber:              input.PhoneNumber,
		DateOfBirth:              input.DateOfBirth,
		Email:                    input.Email,
		Employer:                 input.Employer,
		Occupation:               input.Occupation,
		MailingAddress:           input.MailingAddress,
		PhysicalAddress:          input.PhysicalAddress,
		TrustorIsLiving:          input.TrustorIsLiving,
		DateOfDeath:              input.DateOfDeath,
		TrustorDeathCertReceived: input.TrustorDeathCertReceived,
		TrustorOfSoundMind:       input.TrustorOfSoundMind,
		TrustorHasRelinquished:   input.TrustorHasRelinquished,
		TrustorRelinquishedDate:  input.TrustorRelinquishedDate,
		TrustorRelingDocReceived: input.TrustorRelingDocReceived,
		IsActive:                 1,
		CreatedAt:                time.Now().UTC(),
		UpdatedAt:                time.Now().UTC(),
	}

	// Encrypt SSN if provided
	if input.SSN != nil && *input.SSN != "" {
		// TODO: implement proper encryption
		p.SSNEncrypted = input.SSN
	}

	if err := s.DB.Create(&p).Error; err != nil {
		return nil, err
	}
	p.HasSSN = p.SSNEncrypted != nil && *p.SSNEncrypted != ""
	return &p, nil
}

func (s *PersonService) List(skip, limit int, includeInactive bool, search *string) ([]models.Person, int64, error) {
	var people []models.Person
	var total int64

	q := s.DB.Model(&models.Person{})
	if !includeInactive {
		q = q.Where("is_active = 1")
	}
	if search != nil && *search != "" {
		q = q.Where("name ILIKE ?", "%"+*search+"%")
	}
	q.Count(&total)

	if err := q.Order("name ASC").Offset(skip).Limit(limit).Find(&people).Error; err != nil {
		return nil, 0, err
	}

	for i := range people {
		people[i].HasSSN = people[i].SSNEncrypted != nil && *people[i].SSNEncrypted != ""
	}

	return people, total, nil
}

func (s *PersonService) GetByID(id int) (*models.Person, error) {
	var p models.Person
	if err := s.DB.First(&p, id).Error; err != nil {
		return nil, err
	}
	p.HasSSN = p.SSNEncrypted != nil && *p.SSNEncrypted != ""
	return &p, nil
}

func (s *PersonService) Update(id int, input PersonUpdateInput) (*models.Person, error) {
	var p models.Person
	if err := s.DB.First(&p, id).Error; err != nil {
		return nil, errors.New("person not found")
	}

	if input.Name != nil {
		p.Name = *input.Name
	}
	if input.PhoneNumber != nil {
		p.PhoneNumber = input.PhoneNumber
	}
	if input.DateOfBirth != nil {
		p.DateOfBirth = input.DateOfBirth
	}
	if input.Email != nil {
		p.Email = input.Email
	}
	if input.Employer != nil {
		p.Employer = input.Employer
	}
	if input.Occupation != nil {
		p.Occupation = input.Occupation
	}
	if input.MailingAddress != nil {
		p.MailingAddress = input.MailingAddress
	}
	if input.PhysicalAddress != nil {
		p.PhysicalAddress = input.PhysicalAddress
	}
	if input.TrustorIsLiving != nil {
		p.TrustorIsLiving = input.TrustorIsLiving
	}
	if input.DateOfDeath != nil {
		p.DateOfDeath = input.DateOfDeath
	}
	if input.TrustorDeathCertReceived != nil {
		p.TrustorDeathCertReceived = input.TrustorDeathCertReceived
	}
	if input.TrustorOfSoundMind != nil {
		p.TrustorOfSoundMind = input.TrustorOfSoundMind
	}
	if input.TrustorHasRelinquished != nil {
		p.TrustorHasRelinquished = input.TrustorHasRelinquished
	}
	if input.TrustorRelinquishedDate != nil {
		p.TrustorRelinquishedDate = input.TrustorRelinquishedDate
	}
	if input.TrustorRelingDocReceived != nil {
		p.TrustorRelingDocReceived = input.TrustorRelingDocReceived
	}
	if input.SSN != nil {
		p.SSNEncrypted = input.SSN
	}
	p.UpdatedAt = time.Now().UTC()

	if err := s.DB.Save(&p).Error; err != nil {
		return nil, err
	}
	p.HasSSN = p.SSNEncrypted != nil && *p.SSNEncrypted != ""
	return &p, nil
}

func (s *PersonService) Delete(id int) error {
	var p models.Person
	if err := s.DB.First(&p, id).Error; err != nil {
		return errors.New("person not found")
	}
	p.IsActive = 0
	p.UpdatedAt = time.Now().UTC()
	return s.DB.Save(&p).Error
}

func (s *PersonService) AddRelationship(personID int, input PersonRelationshipInput) error {
	// Verify both people exist
	var p1, p2 models.Person
	if err := s.DB.First(&p1, personID).Error; err != nil {
		return errors.New("person not found")
	}
	if err := s.DB.First(&p2, input.RelatedPersonID).Error; err != nil {
		return errors.New("related person not found")
	}

	rel := models.PersonRelationshipRow{
		PersonID:         personID,
		RelatedPersonID:  input.RelatedPersonID,
		RelationshipType: input.RelationshipType,
		CreatedAt:        time.Now().UTC(),
		UpdatedAt:        time.Now().UTC(),
	}
	return s.DB.Create(&rel).Error
}

func (s *PersonService) RemoveRelationship(personID, relatedPersonID int) error {
	result := s.DB.Where("person_id = ? AND related_person_id = ?", personID, relatedPersonID).
		Delete(&models.PersonRelationshipRow{})
	if result.RowsAffected == 0 {
		return errors.New("relationship not found")
	}
	return result.Error
}

type PersonRelationshipResponse struct {
	PersonID         int            `json:"person_id"`
	RelatedPersonID  int            `json:"related_person_id"`
	RelationshipType string         `json:"relationship_type"`
	RelatedPerson    *models.Person `json:"related_person,omitempty"`
}

func (s *PersonService) GetRelationships(personID int) ([]PersonRelationshipResponse, error) {
	var rels []models.PersonRelationshipRow
	s.DB.Where("person_id = ?", personID).Find(&rels)

	var result []PersonRelationshipResponse
	for _, rel := range rels {
		resp := PersonRelationshipResponse{
			PersonID:         rel.PersonID,
			RelatedPersonID:  rel.RelatedPersonID,
			RelationshipType: rel.RelationshipType,
		}
		var related models.Person
		if err := s.DB.First(&related, rel.RelatedPersonID).Error; err == nil {
			related.HasSSN = related.SSNEncrypted != nil && *related.SSNEncrypted != ""
			resp.RelatedPerson = &related
		}
		result = append(result, resp)
	}
	return result, nil
}
